lib/services/suite/suite.go

```diff
@@ -1011,6 +1013,124 @@ func (s *ServicesTestSuite) ClusterConfig(c *check.C, opts ...SuiteOption) {
+		Expiry:  time.Hour,
+		Params: services.AcquireSemaphoreParams{
+			SemaphoreKind: services.SemaphoreKindSessionControl,
+			SemaphoreName: "alice",
+			MaxLeases:     locks,
+		},
+	}
+	// we leak lock handles in the spawned goroutines, so
+	// context-based cancellation is needed to cleanup the
+	// background keepalive activity.
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	var wg sync.WaitGroup
+	for i := int64(0); i < locks; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			_, err := services.AcquireSemaphoreLock(ctx, cfg)
+			c.Assert(err, check.IsNil)
+		}()
+	}
+	wg.Wait()
+}
+
+// SemaphoreConcurrency verifies that a large number of concurrent
+// acquisitions result in the correct number of successful acquisitions.
+func (s *ServicesTestSuite) SemaphoreConcurrency(c *check.C) {
+	const maxLeases int64 = 2
+	const attempts int64 = 20
+	cfg := services.SemaphoreLockConfig{
+		Service: s.PresenceS,
+		Expiry:  time.Hour,
+		Params: services.AcquireSemaphoreParams{
+			SemaphoreKind: services.SemaphoreKindSessionControl,
+			SemaphoreName: "alice",
+			MaxLeases:     maxLeases,
+		},
+	}
+	// we leak lock handles in the spawned goroutines, so
+	// context-based cancellation is needed to cleanup the
+	// background keepalive activity.
+	ctx, cancel := context.WithCancel(context.TODO())
+	defer cancel()
+	var success int64
+	var failure int64
+	var wg sync.WaitGroup
+	for i := int64(0); i < attempts; i++ {
+		wg.Add(1)
+		go func() {
+			_, err := services.AcquireSemaphoreLock(ctx, cfg)
+			if err == nil {
+				atomic.AddInt64(&success, 1)
+			} else {
+				atomic.AddInt64(&failure, 1)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+	c.Assert(atomic.LoadInt64(&success), check.Equals, maxLeases)
+	c.Assert(atomic.LoadInt64(&failure), check.Equals, attempts-maxLeases)
+}
+
+// SemaphoreLock verifies correct functionality of the basic
+// semaphore lock scenarios.
+func (s *ServicesTestSuite) SemaphoreLock(c *check.C) {
+	cfg := services.SemaphoreLockConfig{
+		Service: s.PresenceS,
+		Expiry:  time.Hour,
+		Params: services.AcquireSemaphoreParams{
+			SemaphoreKind: services.SemaphoreKindSessionControl,
+			SemaphoreName: "alice",
+			MaxLeases:     1,
+		},
+	}
+	lock, err := services.AcquireSemaphoreLock(context.TODO(), cfg)
+	c.Assert(err, check.IsNil)
+
+	// MaxLeases is 1, so second acquire op fails.
+	_, err = services.AcquireSemaphoreLock(context.TODO(), cfg)
+	fixtures.ExpectLimitExceeded(c, err)
+
+	// Lock is successfully released.
+	lock.Cancel()
+	c.Assert(lock.Error(), check.IsNil)
+
+	// Acquire new lock with super-short expiry
+	cfg.Expiry = time.Millisecond * 750
+	lock, err = services.AcquireSemaphoreLock(context.TODO(), cfg)
+	c.Assert(err, check.IsNil)
+
+	select {
+	case <-lock.Done():
+		c.Fatalf("Unexpected lock failure: %v", lock.Error())
+	case <-time.After(time.Millisecond * 1500):
+		c.Fatalf("timeout waiting for successful lock renewal")
+	case <-lock.Renewed():
+	}
```

Comment:
wrap this in a loop and make sure renewal works >1 time
you may need to lower `cfg.Expiry` to ~100ms